rocAL/include/pipeline/argument.h

```diff
@@ -0,0 +1,261 @@
+            values.push_back(static_cast<int>(val));
+        } else {
+            THROW("Unknown enum type for argument " + arg_name);
+        }
+    }
+
+    /**
+     * @brief Constructs argument from vector type
+     * @tparam T The vector type
+     * @param val The vector value
+     */
+    template<typename T>
+    void constructFromVector(T&& val) {
+        static_assert(is_vector_type_v<std::decay_t<T>>, "T must be a vector type");
+        
+        using ElementType = typename std::decay_t<T>::value_type;
+        const std::string element_type_name = getTypeName<ElementType>();
+        
+        if (element_type_name != "unknown") {
+            type_name = element_type_name;
+            is_vector = true;
+            values.reserve(val.size());
+            
+            auto&& local_val = std::forward<T>(val);
+            for (auto&& v : local_val) {
+                values.push_back(static_cast<ElementType>(std::forward<decltype(v)>(v)));
+            }
+        } else {
+            THROW("Unknown vector element type for argument " + arg_name);
+        }
+    }
+
+    /**
+     * @brief Constructs argument from basic type
+     * @tparam T The basic type
+     * @param val The value
+     */
+    template<typename T>
+    void constructFromBasicType(T&& val) {
+        type_name = getTypeName<T>();
+        
+        if (type_name != "unknown") {
+            if constexpr (std::is_same_v<std::decay_t<T>, const char*>) {
+                values.push_back(std::string(val));
+            } else {
+                values.push_back(static_cast<std::decay_t<T>>(std::forward<T>(val)));
+            }
+        } else {
+            THROW("Unknown type " + std::string(typeid(T).name()) + " for argument " + arg_name);
+        }
+    }
+
+    /**
+     * @brief Constructs argument from shared pointer type
+     * @tparam T The shared_ptr type
+     * @param val The shared_ptr value
+     */
+    template<typename T>
+    void constructFromSharedPtr(T&& val) {
+        static_assert(is_shared_ptr_v<std::decay_t<T>>, "T must be a shared_ptr type");
+        
+        type_name = "shared_ptr";
+        // For MetaDataReader, mark as an external reference to be resolved during deserialization.
+        // The actual MetaDataReader instance should be created and provided by the MasterGraph/Pipeline.
+        if (arg_name == "meta_data_reader") {
+            sub_type_name = "MetaDataReader";
+            // No serialized payload for external references.
+        } else {
+            THROW("Unsupported shared_ptr type for argument " + arg_name);
+        }
+    }
+
+    /**
+     * @brief Constructs argument from string-to-string map type
+     * @tparam T The map type
+     * @param val The map value
+     */
+    template<typename T>
+    void constructFromMap(T&& val) {
+        static_assert(is_string_map_v<std::decay_t<T>>, "T must be a string-to-string map type");
+        
+        type_name = "map_string";
```

Comment:
Please check if this is used somewhere. This gets overwritten in each function.

Reply:
Yes the type name is required, which decides the type of the argument.

Only one of these functions will be called for each argument based on the type of the argument, hence it is overriden